Fixed lazy usage with Suspense and Error Boundary together

kamikazePT opened the pull request:
## Summary

When using lazy ( the Suspense version of loadable ), if any error occured, the Suspense would keep rendering It's fallback and retrying and never being catched by the parent ErrorBoundary.

For example I have a dynamic Icon that has specific components, and depending on the Icon type, it loads the correct component, displaying a default icon if no component is found, the way to do this without an Error Boundary would be to add the catch clause on the import( ) statement.

## Test plan

With this PR, if the import( ) fails because the file doesn't exist, it will then render the Fallback provided to the ErrorBoundary

![Screenshot from 2020-02-04 21-07-39](https://user-images.githubusercontent.com/6997763/73787138-8fb4a780-4792-11ea-80f3-c7c40d3bf31d.png)

kamikazePT commented:
In the meantime I've found another issue regarding Suspense usage.

If I have a full dynamic import aka something like
` lazy(props => import(`./${props.name}`))`

and "name" is changing runtime, we would be blocked on a loop on the throw loadAsync call

kamikazePT commented:
@gregberge yeah... but caching the promise, how would you get the result in a sync way to change this piece of code inside the render method? to cache the promise instead of the result/error the render method needs to be re-thinked, no?

```
      if (options.suspense && !this.promise) {
          const cachedResponse = this.getCache()
          if (!cachedResponse) throw this.loadAsync()
          if (cachedResponse.error) throw cachedResponse.error

          return render({
            loading: false,
            fallback: null,
            result: cachedResponse.result,
            options,
            props: { ...props, ref: forwardedRef },
          })
        }

```

kamikazePT commented:
@gregberge 
pushed changes :)

tests are passing, promise cached and render method looks cleaner

theKashey commented:
There is another PR, which fixes the same problem - https://github.com/gregberge/loadable-components/pull/516

kamikazePT commented:
Actually I think that PR is somewhat a complement to this one, he doesnt fix the error not being propagated to the error boundary, he fix's the re-fetching on rehydration after ssr

Atleast that was what I understood from the change he made which was on loadSync that runs on ssr

Still I dont believe that the cache persists from ssr to csr o.O which makes his fix do nothing

theKashey commented on packages/component/src/createLoadable.js:
```diff
@@ -184,15 +187,8 @@ function createLoadable({ resolve = identity, render, onLoad }) {
-            props: { ...props, ref: forwardedRef },
-          })
+          const cachedPromise = this.getCache()
+          if (!cachedPromise) throw this.loadAsync()
```
That would happen if you have two components? What if this component would be rendered twice?
In both cases the second one would be  "cached" and probably would not work.

Probably the cache should hold information about the status of a cached item.

kamikazePT replied:
If you have 2 components at the same tree level, they should have different Keys (like in arrays), having different Keys makes them different instances with their own internal cache

What are your thoughts on that?

theKashey replied:
You might code split  absolutely everything, including a single `Button`, you can use 20 times on a single page.
So - it could be the same cache key, repeated 20 times.

kamikazePT replied:
I've added some tests for multiple rendered components
Please check them out to see if I understood you correctly

theKashey commented on packages/component/src/createLoadable.js:
```diff
@@ -151,16 +151,19 @@ function createLoadable({ resolve = identity, render, onLoad }) {
+
+          const cachedPromise = this.getCache()
+
+          this.promise = cachedPromise || ctor.requireAsync(props)
```
What if the cache is a bit obsolete?
What if it was rejected by any reason? 

Now you are not able to restart the operation.

kamikazePT replied:
I would make the promise retryable, instead of re-rendering from the error boundary :) although there might be use cases for the re-render, I'll think for a bit about this

theKashey replied:
Promise cannot be retryable once it resolved or rejected - those are final states.
And in case of some error the normal expectation nowadays is to throw that error, and handle it at the nearest boundary.

kamikazePT replied:
this was what I had in mind when I said "making" the promise retryable
```
retry = (fn, retries = 3) =>
    fn().catch(err => 
            retries > 1 
                ? retry(fn, retries - 1) 
                : Promise.reject(err))

loadable(() => retry(() => import('foo')))
```

theKashey replied:
It's good to have "retry" working out of the box, but usually, it's better to expose this ability to the user, as long as they might want to communicate the problem to the end-user.

theKashey commented on packages/component/src/loadable.test.js:
```diff
@@ -242,6 +279,21 @@ describe('#lazy.lib', () => {
+    )
+    expect(container).toHaveTextContent('progress')
+    load.reject(new Error('Error Boundary'))
+    await wait(() => expect(container).toHaveTextContent('error'))
```
Please add one more test - `supports retry from the Error Boundary`. It shall not be a final, unrecoverable, state.

gregberge commented:
Hello @kamikazePT, thank you for trying, but I think the approach is not the good one. The best one is to create a container that encapsulate the promise.

An example:

```js
export const PENDING = 0
export const RESOLVED = 1
export const REJECTED = 2

function createContainer(status, result) {
  const container = {
    status,
    result,
    write(result) {
      if (container.status !== RESOLVED) {
        throw new Error('Cannot write container if not resolved')
      }
      return { ...container, result, toJSON: () => result }
    },
    toJSON() {
      return result
    },
  }
  return container
}

export function createContainerFromPromise(promise) {
  const container = createContainer(PENDING)
  container.promise = promise.then(
    r => {
      container.status = RESOLVED
      container.result = r
      container.toJSON = () => r
    },
    e => {
      container.status = REJECTED
      container.error = e
    },
  )
  return container
}

export function createContainerFromValue(value) {
  return createContainer(RESOLVED, value)
}
```

kamikazePT commented:
@gregberge 
@theKashey 

So with some hammer magic I managed to make retrying work for both lazy and loadable variants

It needs some code cleaning but atleast it's tested and working at the moment...

theKashey commented on packages/component/src/util.js:
```diff
@@ -12,3 +12,33 @@ export function warn(message) {
+export function statusAware(promise) {
+  let status = STATUS_PENDING
+
+  const proxy = new Proxy(promise, {
```
No `Proxies` please. They are not supported widely enough yet.
And why this needs a recursive wrapping?

kamikazePT replied:
The recursion is that if you call then or catch, it enables you to access status on that newly created promises from those calls

theKashey commented on packages/component/src/createLoadable.js:
```diff
@@ -90,7 +90,17 @@ function createLoadable({ resolve = identity, render, onLoad }) {
+            loading: true,
+          })
+
+          if (!options.suspense) this.loadAsync()
```
You can load it async with suspense as well, why not?

kamikazePT replied:
Because you need to throw the loadAsync call on the render method so that Suspense catches it, doing that on DidMount makes you add a lot more if conditions at the render method to support more corner cases....sometimes the Best Solution to a problem is to avoid that problem :/

kamikazePT replied:
@theKashey 
Besides adding a fallback for Proxy, is there anything else preventing the acceptance of this PR?

kamikazePT commented:
@theKashey 
@gregberge 

Besides adding a fallback for Proxy usage when not available, is there anything else that would block your acceptance?
I have been busy but I'll proceed with that change in a few days

gregberge commented:
@kamikazePT I am sorry but i don't have time to dig into this right now.

kamikazePT commented:
using a fork with this PR in a production app, just noticed I added a bug when fixed the retrying functionality

also tests weren't exactly bulletproof and could lead to false assumptions, they were improved as well

theKashey commented:
Let's double-check what's missing.

kamikazePT commented:
@theKashey 
PR updated with latest commits on master

theKashey commented:
👀

kamikazePT commented:
@theKashey 

revisited the PR, removed the ES6 Proxy

kamikazePT commented:
@theKashey 
@gregberge 

Anything you need to discuss regarding the solution? or the refactor in the unit tests mocks of the load function?

theKashey commented:
Hey @kamikazePT, sorry for being so slow. Just two moments left:
- [ ] `::` operator used in test (is it "babel-bind"? Is there any way not to use it?)
- [ ] adding callbacks to pending promises

kamikazePT commented:
@theKashey 

the bind operator is only there for sugar syntax and since it's only being used in testing environment, it won't add overhead to the bundling / transpiling for the release files,  in my honest opinion I would keep it :)

can you explain a bit more what you mean with adding callbacks for pending promises?

theKashey commented:
> can you explain a bit more what you mean with adding callbacks for pending promises?

Every time you call `.loadAsync` it does `.then` It is possible to call it multiple times, due to multiple reasons, and it will "stack" callbacks, while only one is required.

kamikazePT commented:
@theKashey 

Fixed the callback situation and removed unnecessary bits of code that weren't doing anything (tests are passing)

Please check it again :)
Hopefully everything is sorted!

theKashey commented:
Very good @kamikazePT 👍

theKashey commented:
PS: 🥳
